fix(changelog): fix missing commit fields in context (#837)

dqkqd opened the pull request:
## Description

Add new fields `raw_message` in `Commit` to avoid lossy message conversion.

Fix #837 

## Motivation and Context
I implemented according to https://github.com/orhun/git-cliff/issues/837#issuecomment-2404818161
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Add 2 test fixtures
- Create changelog with all fields from conventional commits
- Run `git cliff --context | git cliff --from-context` should yield identical result with `git cliff`

## Screenshots / Logs (if applicable)

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- Thank you for contributing to git-cliff! ⛰️  -->

dqkqd commented on git-cliff-core/src/commit.rs:
```diff
@@ -423,8 +433,9 @@ impl Serialize for Commit<'_> {
 		}
 
-		let mut commit = serializer.serialize_struct("Commit", 9)?;
+		let mut commit = serializer.serialize_struct("Commit", 20)?;
```
I counted there are 20 fields, so I put these as 20. Not sure if it is correct.
I found in `cargo` they use separated struct for serializing. Maybe we should do that as well?
https://github.com/rust-lang/cargo/blob/82c489f1c612096c2dffc48a4091d21af4b8e046/src/cargo/core/manifest.rs#L324

orhun replied:
Yup, 20 should be correct.

>  they use separated struct for serializing.

I'm not sure. If you think it will simplify the code, yes. But the current serialization is already super straightforward :)

What would be the benefit of using a separate struct?

dqkqd replied:
> What would be the benefit of using a separate struct?

Actually, I'm pretty new to serde so I'm not quite sure.
But, I think we don't need to manually change the number of fields when using separated struct.
We can also use proc macro directly, which is more friendly.

But those fields aren't change quite often, so I think just leave them there is also good.

orhun commented on git-cliff-core/src/commit.rs:
```diff
@@ -476,6 +487,24 @@ impl Serialize for Commit<'_> {
 	}
 }
 
+/// Deserialize commits into conventional commits if they are convertible
```
Nit:
```suggestion
/// Deserialize commits into conventional commits if they are convertible.
```

dqkqd replied:
> What would be the benefit of using a separate struct?

Actually, I'm pretty new to serde so I'm not quite sure.
But, I think we don't need to manually change the number of fields when using separated struct.
We can also use proc macro directly, which is more friendly.

But those fields aren't change quite often, so I think just leave them there is also good.

orhun commented on .github/workflows/test-fixtures.yml:
```diff
@@ -93,6 +94,8 @@ jobs:
           - fixtures-name: test-from-context
             command: --from-context context.json
+          - fixtures-name: test-from-context-does-not-discard-fields
+            command: --context | git cliff --from-context -
```
I like this, but I think here we are not comparing the final output against the `git cliff --context`, instead we are asserting what we get from `--from-context`.

It would be better if we add a `context.json` file to the fixture and use that instead of running `--context`. What do you think?

dqkqd replied:
> It would be better if we add a context.json file to the fixture and use that instead of running --context. What do you think?

Yes that's great, it would make the test case deterministic instead of depending on output from `--context`. I will fix that.

orhun reviewed (commented):
Neat!

Left some comments :) Also, can you update the documentation (in `/website`) about the existence of the new `raw_message` field?

dqkqd commented:
@orhun

I fixed according to your comments.
I also added the new field to [context.md](https://github.com/orhun/git-cliff/blob/main/website/docs/templating/context.md), but I'm not sure is there any other places I should look into.

orhun commented on .github/fixtures/test-from-context-does-not-discard-fields/commit.sh:
These test commits should no longer be needed. Can you test with an empty `commit.sh` and commit it if it works?

P.S. we should do the same for `test-from-context` fixture too.

dqkqd replied:
Sorry, it didn't work. I got an error, we must at least have one commit. I think this is a bug, but maybe we should fix this in another PR?

```bash
ERROR git_cliff > Git error: `reference 'refs/heads/main' not found; class=Reference (4); code=UnbornBranch (-9)`
```

However, even if we don't use `commit.sh`, I think we should just keep it there, as a reference. I find it more difficult to read `context.json` compared to `commit.sh`

orhun replied:
Ah, okay. Yup, we need to fix that bug in another PR.

> I find it more difficult to read context.json compared to commit.sh

Fair enough!

orhun commented on check.md:
Accidentally committed? 🙂

dqkqd replied:
My bad :(, removed it

orhun reviewed (commented):
One last thing :)

orhun reviewed (approved):
This is great, thanks for your contribution!

janbuchar commented:
Hi @orhun, I just encountered this issue. Could you please make a release that includes this fix?

orhun commented:
Yup, I will try to do a release next week.

janbuchar commented:
> Yup, I will try to do a release next week.

Sorry to bother you, but any chance you could release it?

orhun commented:
released now :)
